src/main/java/org/logstash/dissect/Dissector.java

```diff
@@ -170,56 +164,37 @@ private int dissectValues(byte[] source, Map<Field, ValueRef> fieldValueRefMap)
                     repeatedDelimiters = false;
                 }
             }
-            while (!found) {
-                // each delimiter is given a strategy that the indexOf method
-                // uses to search in the source bytes for itself starting from
-                // where we think next field might begin (left)
-                pos = next.indexOf(source, left);
-                if (pos == left) {
-                    // we found a delimiter
-                    left = pos + next.size();
-                } else {
-                    found = true;
-                }
-            }
+            fieldStart = left;
+            final Delimiter next = field.nextDelimiter();
+            fieldLength = 0;
+            pos = next.indexOf(source, left);
             if (pos > 0) {
-                // now we have found a field and pos is advanced to the next delimiter
-                // remap the field to a proper ValueRef.
-                fieldValueRefMap.put(field, new ValueRef(left, pos - left));
-                // set left to be the end of the delimiter - the start index of the next field
+                // pos is at the next delimiter
+                // we have found the end of the field
+                fieldLength = pos - left;
+                // set left to be the end of the delimiter
+                // where we hope is the start index of the next field
                 left = pos + next.size();
             }
+            fieldValueRefs[field.id()].update(fieldStart, fieldLength);
         }
-        boolean repeatedLastDelimiters = (pos > 0);
-        Delimiter prev = lastField.previousDelimiter();
-        while (repeatedLastDelimiters) {
+        final Field lastField = fields.get(numFields - 1);
+        fieldValueRefs[lastField.id()].clear();
+        prev = lastField.previousDelimiter();
+        // the user indicated greedy with -> OR the first field had delimiter(s) before it
+        boolean repeatedDelimiters = prev != null && prev.isGreedy();
+        while (repeatedDelimiters) {
             pos = prev.indexOf(source, left);
             if (pos == left) { // we found a delimiter
                 left = pos + prev.size();
             } else {
-                repeatedLastDelimiters = false;
+                repeatedDelimiters = false;
```

Comment:
I think its a matter of style, but we could break instead of using a flag?

Reply:
Will look into it.